### ISSUE TITLE
syscall: ztypes_linux_s390x.go: struct members with just underscore as names

### ISSUE THREAD

**Issue by sharkcz** [golang/go#18628]
Structure members in syscall/ztypes_linux_s390x.go that should have a underscore in their names contain only the underscore as the name. It results in error like

<pre>
# github.com/vishvananda/netlink/nl
../../rpmbuild/BUILDROOT/golang-github-vishvananda-netlink-0-0.10.gite73bad4.fc26.s390x/usr/share/gocode/src/github.com/vishvananda/netlink/nl/nl_linux_test.go:34: msg.X__ifi_pad undefined (type *IfInfomsg has no field or method X__ifi_pad)
FAIL	github.com/vishvananda/netlink/nl [build failed]
</pre>

For a downstream bug please see https://bugzilla.redhat.com/show_bug.cgi?id=1404679

https://github.com/golang/sys/blob/master/unix/ztypes_linux_s390x.go from the standalone sys package is wrong as well.

### What version of Go are you using (`go version`)?
git

### What operating system and processor architecture are you using (`go env`)?
Linux/s390x

### What did you do?
compile eg. golang-github-vishvananda-netlink library

### What did you expect to see?
<pre>
type IfInfomsg struct {
		Family     uint8
		X__ifi_pad uint8
		Type       uint16
		Index      int32
		Flags      uint32
		Change     uint32
	}
</pre>

for example on https://golang.org/src/syscall/ztypes_linux_ppc64.go 

The corresponding kernel structure is defined in include/uapi/linux/rtnetlink.h as 
<pre>
struct ifinfomsg {
        unsigned char   ifi_family;
        unsigned char   __ifi_pad;
        unsigned short  ifi_type;               /* ARPHRD_* */
        int             ifi_index;              /* Link index   */
        unsigned        ifi_flags;              /* IFF_* flags  */
        unsigned        ifi_change;             /* IFF_* change mask */
};
</pre>

### What did you see instead?
from https://golang.org/src/syscall/ztypes_linux_s390x.go
<pre>
type IfInfomsg struct {
		Family uint8
		_      uint8
		Type   uint16
		Index  int32
		Flags  uint32
		Change uint32
    }
</pre>

**Comment by sharkcz** [comment]
The IfInfomsg structure is only a sample, there are more with this error.

**Comment by randall77** [comment]
This looks intentional.  Field names that start with `X_` are replaced with `_` by src/syscall/mkpost.go:42-43.

It is just a padding field - why do you need to refer to it?

**Comment by mundaym** [comment]
The fields were omitted from the API exposed on s390x because they weren't considered to have any valid uses. If that assumption is wrong then I'm happy to expose them in the x/sys/unix package at least (the syscall package is supposed to be frozen and shouldn't be used).

The `__ifi_pad` is just padding. It isn't in the man page for rtnetlink. There the struct is described as:

```c
struct ifinfomsg {
    unsigned char  ifi_family; /* AF_UNSPEC */
    unsigned short ifi_type;   /* Device type */
    int            ifi_index;  /* Interface index */
    unsigned int   ifi_flags;  /* Device flags  */
    unsigned int   ifi_change; /* change mask */
};
```

In my opinion the netlink library should therefore not rely on its presence. Unfortunately it is not just a one line change to do this, the tests will have to be rethought a little as the padding field appears not to be cleared, so the equality checks fail (here is the code: https://github.com/vishvananda/netlink/blob/90b9ee53585fd2ab8360c6bde4c9f19f23169ac8/nl/nl_linux_test.go#L34).

**Comment by sharkcz** [comment]
We found at least 2 projects that are touching the padding fields, one is the mentioned netlink, another appeared right today

# github.com/pkg/sftp
../../BUILDROOT/golang-github-pkg-sftp-0-0.1.git8197a2e.fc26.s390x/usr/share/gocode/src/github.com/pkg/sftp/server_statvfs_linux.go:19: stat.Fsid.X__val undefined (type syscall.Fsid has no field or method X__val)
FAIL	github.com/pkg/sftp [build failed]

So my preference would be to add proper names also to the padding fields, because people do weird things :-)

**Comment by sharkcz** [comment]
Also I'm not sure if only the paddings are named with underscore or some valid members are misnamed. But a consistent solution across all arches would be appreciated.

**Comment by bradfitz** [comment]
@sharkcz, that project was already fixed on Sep 30th 2016 via https://github.com/pkg/sftp/commit/4d0e916071f68db74f8a73926335f809396d6b42 for duplicate bug #17298

**Comment by sharkcz** [comment]
Ah, thanks for, I'll ask our maintainers to update the sftp package. Dunno why they used older snapshot, although the package was introduced into Fedora yesterday.

**Comment by bradfitz** [comment]
The meta issue is #18632.

**Comment by tklauser** [comment]
The originally reported issue seems to be have been fixed in vishvananda/netlink@a1d6c19f199e20aba37d4322b5d7bf85dc14fae3 as well.

Can we close this issue then?

**Comment by ianlancetaylor** [comment]
I think so.  Closing.
